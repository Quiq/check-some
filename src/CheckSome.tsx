import React, {createContext} from 'react';
import isEqual from 'lodash/isEqual';
import CheckSomeField from './Field';
import {ValidationErrors} from './globals';

type ValidationRule<T> = (value: T) => ValidationErrors | null;

type ValidationGroupValues<T> = {[K in keyof T]: T[K]};
type ValidationGroupRules<T> = Partial<{[K in keyof T]: Array<ValidationRule<T[K]>>}>;
type ValidationGroupErrors<T> = Partial<{[K in keyof T]: ValidationErrors}> | null;

export type CheckSomeChildProps<T> = {
  valid: boolean;
  changed: boolean;
  errors: ValidationGroupErrors<T>;
};

export type CheckSomeProps<T> = {
  rules: ValidationGroupRules<T>;
  values: ValidationGroupValues<T>;
  initialValues?: ValidationGroupValues<T>;
  children: (props: CheckSomeChildProps<T>) => React.ReactNode;
};

export const CheckSomeContext = createContext({});

export default class CheckSome<T> extends React.Component<CheckSomeProps<T>> {
  static Field = CheckSomeField;

  static defaultProps = {
    rules: {},
  };

  initialValues: Object | null | undefined;

  getInitialValues = () => {
    if (this.props.initialValues) {
      return this.props.initialValues;
    }

    if (!this.initialValues) {
      this.initialValues = this.props.values;
    }

    return this.initialValues;
  };

  getErrors = (): ValidationGroupErrors<T> =>
    Object.keys(this.props.rules).reduce((errors: ValidationGroupErrors<T>, keyValue) => {
      // @ts-ignore
      const key: keyof T = keyValue;
      const rules = this.props.rules[key];
      const value = this.props.values[key];

      const newErrors = rules!.reduce(
        (e: ValidationErrors | null, rule: ValidationRule<T[typeof key]>) => e || rule(value),
        null,
      );

      if (!newErrors) {
        return errors;
      }
      if (!errors) {
        // Initialize error object if this is the first error
        errors = {};
      }

      errors[key] = newErrors;

      return errors;
    }, null);

  render() {
    const {values} = this.props;

    const errors = this.getErrors();
    const valid = !errors;

    const changed = !isEqual(values, this.getInitialValues());

    return (
      <CheckSomeContext.Provider value={{values, errors}}>
        {this.props.children({valid, errors, changed})}
      </CheckSomeContext.Provider>
    );
  }
}

const greaterThanZero = (value: number) => {
  return value > 0 ? null : {greaterThanZero: {value}};
};

const startsWith = (startingText: string) => (value: string) =>
  value.startsWith(startingText) ? null : {startsWith: {startingText, value}};

const tsCheck = (
  <CheckSome
    values={{
      stringField: 'string',
      numberField: 7,
    }}
    initialValues={{
      stringField: 'string',
      numberField: 7,
    }}
    rules={{
      stringField: [startsWith('test')],
      numberField: [greaterThanZero],
    }}
  >
    {() => <div />}
  </CheckSome>
);
