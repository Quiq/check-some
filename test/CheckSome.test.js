// @flow
import * as React from 'react';
import CheckSome from '../src';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import type {ReactWrapper} from 'enzyme';

Enzyme.configure({adapter: new Adapter()});

const required = value => (value || value === 0 ? null : {required: {}});

const greaterThanZero = (value: number) => (value > 0 ? null : {greaterThanZero: {value}});

const TestField = ({value, errors, valid, touched}) => (
  <div className="field">
    <input defaultValue={value} />
    <div>Field {valid ? 'Valid' : 'Invalid'}</div>
    <div>{touched ? 'Touched' : 'Pristine'}</div>
    <div>Field Errors: {JSON.stringify(errors)}</div>
  </div>
);

const TestForm = ({valid, changed, errors}) => (
  <form>
    <div>Form {valid ? 'Valid' : 'Invalid'}</div>
    <div>{changed ? 'Changed' : 'Unchanged'}</div>
    <div>Form Errors: {JSON.stringify(errors)}</div>

    <CheckSome.Field name="requiredString">
      {fieldProps => <TestField {...fieldProps} />}
    </CheckSome.Field>

    <CheckSome.Field name="testNumber">
      {fieldProps => <TestField {...fieldProps} />}
    </CheckSome.Field>

    <CheckSome.Field name="optionalString">
      {fieldProps => <TestField {...fieldProps} />}
    </CheckSome.Field>
  </form>
);

describe('CheckSome', () => {
  let wrapper: ReactWrapper;

  beforeEach(() => {
    const values = {
      requiredString: '',
      testNumber: undefined,
      optionalString: '',
    };

    const rules = {
      requiredString: [required],
      testNumber: [required, greaterThanZero],
    };

    wrapper = mount(
      <CheckSome values={values} rules={rules}>
        {formProps => <TestForm {...formProps} />}
      </CheckSome>,
    );
  });

  it('renders the initial form', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
