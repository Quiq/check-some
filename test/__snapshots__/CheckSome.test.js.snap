// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckSome blurring the fields sets the touched prop for fields 1`] = `
<CheckSome
  rules={
    Object {
      "requiredString": Array [
        [Function],
      ],
      "testNumber": Array [
        [Function],
        [Function],
      ],
    }
  }
  values={
    Object {
      "optionalString": "",
      "requiredString": "",
      "testNumber": undefined,
    }
  }
>
  <TestForm
    changed={false}
    errors={
      Object {
        "requiredString": Object {
          "required": Object {},
        },
        "testNumber": Object {
          "required": Object {},
        },
      }
    }
    valid={false}
  >
    <form>
      <div>
        Form 
        Invalid
      </div>
      <div>
        Unchanged
      </div>
      <div>
        Form Errors: 
        {"requiredString":{"required":{}},"testNumber":{"required":{}}}
      </div>
      <CheckSomeField
        name="requiredString"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            errors={
              Object {
                "required": Object {},
              }
            }
            touched={true}
            valid={false}
            value=""
          >
            <div
              className="field"
            >
              <input
                defaultValue=""
              />
              <div>
                Field 
                Invalid
              </div>
              <div>
                Touched
              </div>
              <div>
                Field Errors: 
                {"required":{}}
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
      <CheckSomeField
        name="testNumber"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            errors={
              Object {
                "required": Object {},
              }
            }
            touched={true}
            valid={false}
          >
            <div
              className="field"
            >
              <input />
              <div>
                Field 
                Invalid
              </div>
              <div>
                Touched
              </div>
              <div>
                Field Errors: 
                {"required":{}}
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
      <CheckSomeField
        name="optionalString"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            touched={true}
            valid={true}
            value=""
          >
            <div
              className="field"
            >
              <input
                defaultValue=""
              />
              <div>
                Field 
                Valid
              </div>
              <div>
                Touched
              </div>
              <div>
                Field Errors: 
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
    </form>
  </TestForm>
</CheckSome>
`;

exports[`CheckSome renders the initial form 1`] = `
<CheckSome
  rules={
    Object {
      "requiredString": Array [
        [Function],
      ],
      "testNumber": Array [
        [Function],
        [Function],
      ],
    }
  }
  values={
    Object {
      "optionalString": "",
      "requiredString": "",
      "testNumber": undefined,
    }
  }
>
  <TestForm
    changed={false}
    errors={
      Object {
        "requiredString": Object {
          "required": Object {},
        },
        "testNumber": Object {
          "required": Object {},
        },
      }
    }
    valid={false}
  >
    <form>
      <div>
        Form 
        Invalid
      </div>
      <div>
        Unchanged
      </div>
      <div>
        Form Errors: 
        {"requiredString":{"required":{}},"testNumber":{"required":{}}}
      </div>
      <CheckSomeField
        name="requiredString"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            errors={
              Object {
                "required": Object {},
              }
            }
            touched={false}
            valid={false}
            value=""
          >
            <div
              className="field"
            >
              <input
                defaultValue=""
              />
              <div>
                Field 
                Invalid
              </div>
              <div>
                Pristine
              </div>
              <div>
                Field Errors: 
                {"required":{}}
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
      <CheckSomeField
        name="testNumber"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            errors={
              Object {
                "required": Object {},
              }
            }
            touched={false}
            valid={false}
          >
            <div
              className="field"
            >
              <input />
              <div>
                Field 
                Invalid
              </div>
              <div>
                Pristine
              </div>
              <div>
                Field Errors: 
                {"required":{}}
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
      <CheckSomeField
        name="optionalString"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            touched={false}
            valid={true}
            value=""
          >
            <div
              className="field"
            >
              <input
                defaultValue=""
              />
              <div>
                Field 
                Valid
              </div>
              <div>
                Pristine
              </div>
              <div>
                Field Errors: 
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
    </form>
  </TestForm>
</CheckSome>
`;

exports[`CheckSome setting values to something still invalid updates changed, values, and errors 1`] = `
<CheckSome
  rules={
    Object {
      "requiredString": Array [
        [Function],
      ],
      "testNumber": Array [
        [Function],
        [Function],
      ],
    }
  }
  values={
    Object {
      "optionalString": "patrick",
      "requiredString": "spongebob",
      "testNumber": -2,
    }
  }
>
  <TestForm
    changed={true}
    errors={
      Object {
        "testNumber": Object {
          "greaterThanZero": Object {
            "value": -2,
          },
        },
      }
    }
    valid={false}
  >
    <form>
      <div>
        Form 
        Invalid
      </div>
      <div>
        Changed
      </div>
      <div>
        Form Errors: 
        {"testNumber":{"greaterThanZero":{"value":-2}}}
      </div>
      <CheckSomeField
        name="requiredString"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            touched={false}
            valid={true}
            value="spongebob"
          >
            <div
              className="field"
            >
              <input
                defaultValue="spongebob"
              />
              <div>
                Field 
                Valid
              </div>
              <div>
                Pristine
              </div>
              <div>
                Field Errors: 
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
      <CheckSomeField
        name="testNumber"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            errors={
              Object {
                "greaterThanZero": Object {
                  "value": -2,
                },
              }
            }
            touched={false}
            valid={false}
            value={-2}
          >
            <div
              className="field"
            >
              <input
                defaultValue={-2}
              />
              <div>
                Field 
                Invalid
              </div>
              <div>
                Pristine
              </div>
              <div>
                Field Errors: 
                {"greaterThanZero":{"value":-2}}
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
      <CheckSomeField
        name="optionalString"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            touched={false}
            valid={true}
            value="patrick"
          >
            <div
              className="field"
            >
              <input
                defaultValue="patrick"
              />
              <div>
                Field 
                Valid
              </div>
              <div>
                Pristine
              </div>
              <div>
                Field Errors: 
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
    </form>
  </TestForm>
</CheckSome>
`;

exports[`CheckSome setting values to something valid updates changed, values, and errors 1`] = `
<CheckSome
  rules={
    Object {
      "requiredString": Array [
        [Function],
      ],
      "testNumber": Array [
        [Function],
        [Function],
      ],
    }
  }
  values={
    Object {
      "optionalString": "patrick",
      "requiredString": "spongebob",
      "testNumber": 7,
    }
  }
>
  <TestForm
    changed={true}
    errors={null}
    valid={true}
  >
    <form>
      <div>
        Form 
        Valid
      </div>
      <div>
        Changed
      </div>
      <div>
        Form Errors: 
        null
      </div>
      <CheckSomeField
        name="requiredString"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            errors={null}
            touched={false}
            valid={true}
            value="spongebob"
          >
            <div
              className="field"
            >
              <input
                defaultValue="spongebob"
              />
              <div>
                Field 
                Valid
              </div>
              <div>
                Pristine
              </div>
              <div>
                Field Errors: 
                null
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
      <CheckSomeField
        name="testNumber"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            errors={null}
            touched={false}
            valid={true}
            value={7}
          >
            <div
              className="field"
            >
              <input
                defaultValue={7}
              />
              <div>
                Field 
                Valid
              </div>
              <div>
                Pristine
              </div>
              <div>
                Field Errors: 
                null
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
      <CheckSomeField
        name="optionalString"
      >
        <div
          className="CheckSomeField"
          onBlur={[Function]}
        >
          <TestField
            errors={null}
            touched={false}
            valid={true}
            value="patrick"
          >
            <div
              className="field"
            >
              <input
                defaultValue="patrick"
              />
              <div>
                Field 
                Valid
              </div>
              <div>
                Pristine
              </div>
              <div>
                Field Errors: 
                null
              </div>
            </div>
          </TestField>
        </div>
      </CheckSomeField>
    </form>
  </TestForm>
</CheckSome>
`;
